ddsrecorder_participants/src/cpp/mcap/McapHandler.cpp

```diff
@@ -0,0 +1,460 @@
+            );
+        }
+        pending_queue.pop();
+    }
+}
+
+void McapHandler::event_thread_routine_()
+{
+    while (paused_)
+    {
+        clear_all_();
+
+        bool timeout;
+        {
+            std::unique_lock<std::mutex> lock(event_cv_mutex_);
+
+            timeout = !event_cv_.wait_for(
+                lock,
+                std::chrono::seconds(event_window_),
+                [&]
+                {
+                    return event_triggered_;
+                });
+            event_triggered_ = false;
+        }
+
+        if (timeout)
+        {
+            logInfo(DDSRECORDER_MCAP_HANDLER, "Event thread timeout.");
+        }
+        else
+        {
+            if (paused_)
+            {
+                logInfo(DDSRECORDER_MCAP_HANDLER, "Event triggered: dumping buffered data.");
+                dump_data_();
+
+            }
+            else
+            {
+                logInfo(DDSRECORDER_MCAP_HANDLER, "Finishing event thread routine.");
+                clear_all_();
+                break;
+            }
+        }
+    }
+}
+
+void McapHandler::stop_event_thread_()
+{
+    logInfo(DDSRECORDER_MCAP_HANDLER, "Stopping event thread.");
+    paused_.store(false);
+    if (event_thread_.joinable())
+    {
+        {
+            std::lock_guard<std::mutex> lock(event_cv_mutex_);
+            event_triggered_ = true;
+        }
+        event_cv_.notify_one();
+        event_thread_.join();
+    }
+}
+
+void McapHandler::clear_all_()
+{
+    logInfo(DDSRECORDER_MCAP_HANDLER, "Cleaning all buffers.");
+
+    {
+        // Clear samples buffer
+        std::lock_guard<std::mutex> guard(write_mtx_);
+        std::queue<Message> empty;
+        std::swap(samples_buffer_, empty);
+    }
+
+    {
+        // Clear pending samples
+        std::lock_guard<PendingSamplesMapType> lock(pending_samples_);
+        pending_samples_.clear();
+    }
+}
+
+void McapHandler::dump_data_()
+{
+    std::lock_guard<std::mutex> guard(write_mtx_);
+    dump_data_nts_();
+}
+
+void McapHandler::dump_data_nts_()
+{
+    logInfo(DDSRECORDER_MCAP_HANDLER, "Writing data stored in buffer.");
+
+    while (!samples_buffer_.empty())
+    {
+        auto& sample = samples_buffer_.front();
+        auto status = mcap_writer_.write(sample);
+        if (!status.ok())
+        {
+            throw utils::InconsistencyException(
+                STR_ENTRY << "Error writting in MCAP"
+            );
+        }
+        samples_buffer_.pop();
+    }
+}
+
+mcap::ChannelId McapHandler::create_channel_id_nts_(
+        const DdsTopic& topic)
+{
+    // Find schema
+    auto schema_id = get_schema_id_(topic.type_name);
+
+    // Create new channel
+    mcap::Channel new_channel(topic.m_topic_name, "cdr", schema_id);
+    mcap_writer_.addChannel(new_channel);
+    auto channel_id = new_channel.id;
+    channels_.insert({topic.m_topic_name, std::move(new_channel)});
```

Comment:
Channels does not seem thread safe. At least it does not use its own mutex

Reply:
Now protected by unique object mutex.